### PR TITLE
Add support for AI21 models

### PR DESCRIPTION
**Pull request by alexisrobert**
Hello :)

I added support for AI21.com LLM models in a new provider. This is my first contribution for Dust, so I hope my patch is clean :)

(I also added a small edit for the Cohere provider where there's a small typo where the return provider arg was OpenAI instead of Cohere)

### CODE REVIEW

**Review by spolu — COMMENTED** [review]
Looks great. Need to test it and left a small question.

Also no support for token streaming?

File: core/src/providers/ai21.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+                    Some(f) => f,
+                    None => 256,
+                },
+                0, // min_tokens
+                temperature,
+                match top_p {
+                    Some(f) => f,
+                    None => 1.0,
+                },
+                stop,
```
[the remark this answers is not in the document]
Comment (alexisrobert) [reply core/src/providers/ai21.rs:201]: You mean topKReturn ? If so, indeed it looks so indeed, I'll submit a new patch with support for it. But I'm not 100% sure the parameters are iso.

**Comment by alexisrobert** [comment]
Thanks :) I haven't seen support for token streaming in the documentation : https://docs.ai21.com/reference/j1-complete — but I might be wrong.

(I've based myself on the Cohere provider code, as their APIs seemed pretty near from each other)

File: front/pages/api/providers/[pId]/check.js [hunk 2]
```diff
@@ -60,6 +60,27 @@ export default async function handler(req, res) {
...
+        case "ai21":
+          let testTokenize = await fetch(
+            "https://api.ai21.com/studio/v1/tokenize",
+            {
+              method: "POST",
+              headers: {
+                Authorization: `Bearer ${config.api_key}`,
+                "Content-Type": "application/json",
+              },
+              body: JSON.stringify({ text: "Hello World" }),
```
Comment (spolu) [inline front/pages/api/providers/[pId]/check.js:72]: Let's add a max tokens = 1

File: front/components/providers/AI21Setup.jsx [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+              leaveTo="opacity-0"
+            >
+              <Dialog.Panel className="relative transform overflow-hidden rounded-lg bg-white px-4 pt-5 pb-4 text-left shadow-xl transition-all sm:my-8 sm:w-full sm:max-w-sm lg:max-w-lg sm:p-6">
+                <div>
+                  <div className="mt-3">
+                    <Dialog.Title
+                      as="h3"
+                      className="text-lg font-medium leading-6 text-gray-900"
+                    >
+                      Setup AI21 Studio
```
Comment (spolu) [inline front/components/providers/AI21Setup.jsx:92]: Should we remove the "Studio" and just say "AI21" here and in the provider name?

**Review by spolu — CHANGES_REQUESTED** [review]

**Comment by spolu** [comment]
Note there is no streaming support here, but I haven't been able to find it available for AI21

File: front/components/providers/AI21Setup.jsx (hunk 3 above)
Comment (spolu) [reply front/components/providers/AI21Setup.jsx:92]: As per IRL: let's keep it

**Comment by romainhuet** [comment]
Thank you for this pull request, @alexisrobert!

I wanted to check it out to test AI21 Studio models, and used this opportunity to add the corresponding frontend integration: I added AI21 to the list of model providers with the component to set the API key, and it’s now possible to select `j1-*` models from an `llm` block in the interface when building an app. I also did a tiny Rust refactor and fixed the empty `top_logprobs` to return `None`.

Let me know if you have any thoughts!

**Comment by alexisrobert** [comment]
Thanks @romainhuet and @spolu !! :)

**Pull request merged**